### PR TITLE
Generate transfers

### PR DESCRIPTION
**Pull request by datanel**

### CODE REVIEW

File: src/collection.rs [hunk 1]
```diff
@@ -133,6 +133,20 @@ impl<T> Collection<T> {
...
 
+    /// Iterates over the `&mut T` of the `Collection`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use navitia_model::collection::*;
+    /// let mut c: Collection<i32> = Collection::new(vec![1, 1, 2, 3, 5, 8]);
+    /// let values: Vec<&mut i32> = c.values_mut().collect();
+    /// assert_eq!(values, &[&1, &1, &2, &3, &5, &8]);
```
Comment (TeXitoi) [inline src/collection.rs]: Please give an example with modification of the collection.

File: src/gtfs/mod.rs [hunk 2]
```diff
@@ -74,6 +75,8 @@ where
     collections.networks = networks;
     collections.companies = companies;
     let (stop_areas, stop_points) = read::read_stops(path, &mut comments, &mut equipments)?;
+    let transfers = read::make_transfers(&stop_points);
+    collections.transfers = transfers;
```
Comment (TeXitoi) [inline src/gtfs/mod.rs]: you can do that in one line.

File: src/gtfs/read.rs [hunk 3]
```diff
@@ -549,6 +549,28 @@ pub fn read_stops<P: AsRef<path::Path>>(
...
 
+pub fn make_transfers(
+    stop_points: &CollectionWithId<objects::StopPoint>,
+) -> Collection<objects::Transfer> {
+    info!("Generating tranfers");
+    let mut transfers = vec![];
+    for stop_point_1 in stop_points.values() {
+        for stop_point_2 in stop_points.values() {
+            let distance = stop_point_1.distance_to(stop_point_2);
+            let transfer_time = (distance / 0.785) as u32;
```
Comment (TeXitoi) [inline src/gtfs/read.rs]: hard coded?

File: src/gtfs/read.rs [hunk 4]
```diff
@@ -549,6 +549,28 @@ pub fn read_stops<P: AsRef<path::Path>>(
...
+            transfers.push(objects::Transfer {
```
Comment (TeXitoi) [inline src/gtfs/read.rs]: You don't want to create all connections, only the ones in a minimal distance.

File: src/gtfs/read.rs [hunk 5]
```diff
@@ -549,6 +549,28 @@ pub fn read_stops<P: AsRef<path::Path>>(
...
+                from_stop_id: stop_point_1.id.clone(),
+                to_stop_id: stop_point_2.id.clone(),
+                min_transfer_time: Some(transfer_time),
+                real_min_transfer_time: Some(transfer_time),
```
Comment (TeXitoi) [inline src/gtfs/read.rs]: These 2 values should not be equal in practice.

File: src/objects.rs [hunk 6]
```diff
@@ -715,6 +715,24 @@ pub struct StopPoint {
     pub equipment_id: Option<String>,
     pub fare_zone_id: Option<String>,
 }
+
+// Mean earth radius from International Union of Geodesy Geophysics
+const EARTH_RADIUS: f64 = 6_371.000_8;
+
+impl StopPoint {
+    /// Calculate the orthodromic distance between 2 stop points
+    pub fn distance_to(&self, other: &Self) -> f64 {
```
Comment (TeXitoi) [inline src/objects.rs]: Better to impl that on the Coord object

File: src/gtfs/mod.rs [hunk 7]
```diff
@@ -74,6 +75,8 @@ where
     collections.networks = networks;
     collections.companies = companies;
     let (stop_areas, stop_points) = read::read_stops(path, &mut comments, &mut equipments)?;
+    let transfers = read::make_transfers(&stop_points);
```
Comment (TeXitoi) [inline src/gtfs/mod.rs]: Not using the transfers.txt file? No params? No option to desactivate?

File: src/gtfs/read.rs [hunk 8]
```diff
@@ -549,6 +549,69 @@ pub fn read_stops<P: AsRef<path::Path>>(
...
+    path: P,
+    stop_points: &CollectionWithId<objects::StopPoint>,
+) -> Result<Collection<objects::Transfer>> {
+    info!("Reading tranfers.txt");
+    let path = path.as_ref().join("transfers.txt");
+    let mut rdr = csv::Reader::from_path(&path).with_context(ctx_from_path!(path))?;
+    let mut transfers = vec![];
+    for transfer in rdr.deserialize() {
+        let transfer: Transfer = transfer.with_context(ctx_from_path!(path))?;
+        let from_stop_point = stop_points.get(&transfer.from_stop_id).unwrap();
```
Comment (datanel) [inline src/gtfs/read.rs]: I will return an error if `from_stop_point` does not exist

File: src/collection.rs [hunk 9]
```diff
@@ -133,6 +133,23 @@ impl<T> Collection<T> {
...
+    /// # Examples
+    ///
+    /// ```
+    /// # use navitia_model::collection::*;
+    /// let mut c: Collection<i32> = Collection::new(vec![1, 1, 2, 3, 5, 8]);
+    /// for elem in c.values_mut() {
+    ///     *elem *= 2;
+    /// }
+    /// let values: Vec<&i32> = c.values().collect();
+    /// assert_eq!(values, &[&2, &2, &4, &6, &10, &16]);
```
Comment (TeXitoi) [inline src/collection.rs]:
```rust
assert_eq!(values, Collection::new(vec![2, 2, 4, 6, 10, 16]));
```

File: src/gtfs/read.rs [hunk 10]
```diff
@@ -549,6 +549,82 @@ pub fn read_stops<P: AsRef<path::Path>>(
     Ok((stopareas, stoppoints))
 }
 
+#[derive(Derivative)]
+#[derivative(Default(bound = ""))]
+#[derive(Deserialize, Debug)]
```
Comment (TeXitoi) [inline src/gtfs/read.rs]:
```rust
#[derive(Deserialize, Debug, Derivative)]
#[derivative(Default)]
```

File: src/gtfs/read.rs [hunk 11]
```diff
@@ -549,6 +549,82 @@ pub fn read_stops<P: AsRef<path::Path>>(
...
+        let (min_transfer_time, real_min_transfer_time) = match transfer.transfer_type {
+            TransferType::Recommended => {
+                let distance = from_stop_point.coord.distance_to(&to_stop_point.coord);
+                let transfer_time = (distance / 0.785) as u32;
+
+                (Some(transfer_time), Some(transfer_time + 2 * 60))
+            }
+            TransferType::Timed => (Some(0), Some(0)),
+            TransferType::WithTransferTime => {
+                (transfer.min_transfer_time, transfer.min_transfer_time)
```
Comment (TeXitoi) [inline src/gtfs/read.rs]: Maybe we can emit a warning is transfer_type is none.

File: src/gtfs/read.rs [hunk 12]
```diff
@@ -549,6 +549,82 @@ pub fn read_stops<P: AsRef<path::Path>>(
...
+            }
+            TransferType::NotPossible => (Some(86400), Some(84600)),
```
Comment (TeXitoi) [inline src/gtfs/read.rs]: Err, real is less than normal! :-P

File: src/gtfs/read.rs [hunk 13]
```diff
@@ -1567,13 +1653,35 @@ mod tests {
...
+            create_file_with_content(&tmp_dir, "stops.txt", stops_content);
+            create_file_with_content(&tmp_dir, "transfers.txt", transfers_content);
+
+            let mut comments: CollectionWithId<Comment> = CollectionWithId::default();
+            let mut equipments = EquipmentList::default();
+            let (_, stop_points) =
+                super::read_stops(tmp_dir.path(), &mut comments, &mut equipments).unwrap();
+
+            let transfers = super::read_transfers(tmp_dir.path(), &stop_points).unwrap();
+            assert_eq!(9, transfers.len());
```
Comment (TeXitoi) [inline src/gtfs/read.rs]: You can test the values of one of each type, please?

File: src/gtfs/read.rs [hunk 14]
```diff
@@ -549,6 +549,87 @@ pub fn read_stops<P: AsRef<path::Path>>(
...
+                "Problem reading {:?}: to_stop_id={:?} not found",
+                path,
+                transfer.to_stop_id
+            )
+        })?;
+
+        let (min_transfer_time, real_min_transfer_time) = match transfer.transfer_type {
+            TransferType::Recommended => {
+                let distance = from_stop_point.coord.distance_to(&to_stop_point.coord);
+                let transfer_time = (distance / 0.785) as u32;
```
Comment (TeXitoi) [inline src/gtfs/read.rs]: That's a m/s speed.

File: src/objects.rs [hunk 15]
```diff
@@ -657,6 +657,25 @@ pub struct Coord {
     pub lat: f64,
 }
 
+// Mean earth radius from International Union of Geodesy Geophysics
+const EARTH_RADIUS: f64 = 6_371.000_8;
```
Comment (TeXitoi) [inline src/objects.rs]: That's kilometer

File: src/objects.rs [hunk 16]
```diff
@@ -657,6 +657,25 @@ pub struct Coord {
...
+
+impl Coord {
+    /// Calculate the orthodromic distance in kilometers
```
Comment (TeXitoi) [inline src/objects.rs]: Better to do it the navitia way: pure metric system, i.e. meter, meter/second for speed, etc.

Comment (TeXitoi) [reply src/objects.rs]: We generally call this formula haversine distance https://docs.rs/geo/0.8.2/geo/algorithm/haversine_distance/trait.HaversineDistance.html

Comment (datanel) [reply src/objects.rs]: I don't see your point. This is the same formula with different radius.
You want me to add rust geo ?

File: src/objects.rs [hunk 17]
```diff
@@ -1099,4 +1100,36 @@ mod tests {
         assert!(de("00:AA:00").is_err());
         assert!(de("00:00:AA").is_err());
     }
+
+    fn nearly_equal(x: f64, y: f64, epsilon: f64) -> bool {
+        let delta = (x - y).abs();
```
Comment (TeXitoi) [inline src/objects.rs]: delta can go in the else case

File: src/gtfs/read.rs [hunk 18]
```diff
@@ -1681,7 +1686,74 @@ mod tests {
...
+                        from_stop_id: "sp:01".to_string(),
+                        to_stop_id: "sp:01".to_string(),
+                        min_transfer_time: Some(0),
+                        real_min_transfer_time: Some(0),
+                        equipment_id: None,
+                    },
+                    &Transfer {
+                        from_stop_id: "sp:01".to_string(),
+                        to_stop_id: "sp:02".to_string(),
+                        min_transfer_time: Some(44793),
```
Comment (TeXitoi) [inline src/gtfs/read.rs]: That's a long transfer!

Comment (datanel) [reply src/gtfs/read.rs]: in ms :)

**Pull request merged**